File: 10 Regression Learning/MNIST logistic.py
import numpy as np
import tensorflow as tf
from sklearn.metrics import confusion_matrix
from datetime import datetime as dt

((X_train, y_train), 
 (X_test, y_test)) = tf.keras.datasets.mnist.load_data()

N_train = len(X_train)                  # 60000 training samples
N_test = len(X_test)                    # 10000 test samples
X_train = X_train.reshape(N_train, -1).astype('float32') / 255.0
X_test = X_test.reshape(N_test, -1).astype('float32') / 255.0

start = dt.now()
from sklearn.linear_model import LogisticRegression

np.random.seed(4002)
clf = LogisticRegression(max_iter=10000)
clf.fit(X_train, y_train)
y_hat_logit = clf.predict(X_test)

# Table of predictions and prediction accuracy
tab = confusion_matrix(y_hat_logit, y_test)     # Confusion matrix
print(tab)
print(sum(tab.diagonal()) / len(y_test))        # Accuracy
print(dt.now() - start)                  # Logistic regression

#%%
start = dt.now()
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression

pca = PCA(n_components=.95)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)
print(f'Total number of components: {pca.n_components_}')

np.random.seed(4002)
clf_pca = LogisticRegression(max_iter=10000)
clf_pca.fit(X_train_pca, y_train)
y_hat_pca = clf_pca.predict(X_test_pca)

# Table of predictions and prediction accuracy
tab = confusion_matrix(y_hat_pca, y_test)       # Confusion matrix
print(tab)
print(sum(tab.diagonal()) / len(y_test))        # Accuracy
print(dt.now() - start)                  # PCA + Logistic regression

File: 10 Regression Learning/MNIST logistic.R
library(keras)        # MNIST dataset
library(nnet)

# Load MNIST dataset
mnist <- dataset_mnist()
X_train <- mnist$train$x / 255
y_train <- mnist$train$y
X_test <- mnist$test$x / 255
y_test <- mnist$test$y

X_train <- matrix(X_train, nrow=length(y_train))
X_test <- matrix(X_test, nrow=length(y_test))

df_train <- data.frame(cbind(X_train, y_train))
names(df_train) <- c(paste0('x_', 1:ncol(X_train)), 'y')
df_test <- data.frame(X_test)
names(df_test) <- paste0('x_', 1:ncol(X_train))

start <- Sys.time()
set.seed(4002)
mnl <- multinom(y~., data=df_train, MaxNWts=1e6, maxit=200, trace=F)
y_hat <- predict(mnl, newdata=df_test, type="class")
table(y_hat, y_test)      # Confusion matrix
mean(y_hat == y_test)     # Accuracy
Sys.time() - start        # Logistic regression with full training set

################################################################################
start <- Sys.time()
pca <- prcomp(X_train, center=TRUE, scale=FALSE)
s <- pca$sdev
t <- sum(s^2)           # Compute total variance
cumvar <- cumsum(s^2/t) # Cumulative sum of proportion of variance
(idx <- which(cumvar >= 0.95)[1])

X_train_pca <- predict(pca, newdata=X_train)[,1:idx]
X_test_pca <- predict(pca, newdata=X_test)[,1:idx]
df_train_pca <- data.frame(cbind(X_train_pca, y_train))
names(df_train_pca) <- c(paste0('x_', 1:idx), 'y')
df_test_pca <- data.frame(X_test_pca)
names(df_test_pca) <- paste0('x_', 1:idx)

set.seed(4002)
mnl_pca <- multinom(y~., data=df_train_pca, MaxNWts=1e6, maxit=200, 
                    trace=F)
y_hat_pca <- predict(mnl_pca, newdata=df_test_pca, type="class")
table(y_hat_pca, y_test)    # Confusion matrix
mean(y_hat_pca == y_test)   # Accuracy
Sys.time() - start          # PCA + Logistic regression